examples/api-samples/src/browser/chat/change-set-chat-agent-contribution.ts

```diff
@@ -0,0 +1,132 @@
+
+    @inject(WorkspaceService)
+    protected readonly workspaceService: WorkspaceService;
+
+    @inject(FileService)
+    protected readonly fileService: FileService;
+
+    @inject(ChangeSetFileElementFactory)
+    protected readonly fileChangeFactory: ChangeSetFileElementFactory;
+
+    override async invoke(request: ChatRequestModelImpl): Promise<void> {
+        const roots = this.workspaceService.tryGetRoots();
+        if (roots.length === 0) {
+            request.response.response.addContent(new MarkdownChatResponseContentImpl(
+                'No workspace is open. For using this chat agent, please open a workspace with at least two files in the root.'
+            ));
+            request.response.complete();
+            return;
+        }
+
+        const root = roots[0];
+        const files = root.children?.filter(child => child.isFile);
+        if (!files || files.length < 3) {
+            request.response.response.addContent(new MarkdownChatResponseContentImpl(
+                'The workspace does not contain any files. For using this chat agent, please add at least two files in the root.'
+            ));
+            request.response.complete();
+            return;
+        }
+
+        const fileToAdd = root.resource.resolve('hello/new-file.txt');
+        const fileToChange = files[Math.floor(Math.random() * files.length)];
+        const fileToDelete = files.filter(file => file.name !== fileToChange.name)[Math.floor(Math.random() * files.length)];
+
+        const chatSessionId = request.session.id;
+        const changeSet = new ChangeSetImpl('My Test Change Set');
+        changeSet.addElement(
+            this.fileChangeFactory({
+                uri: fileToAdd,
+                type: 'add',
+                state: 'pending',
+                targetState: 'Hello World!',
+                changeSet,
+                chatSessionId
+            })
+        );
+        if (fileToChange && fileToChange.resource) {
+            changeSet.addElement(
+                this.fileChangeFactory({
+                    uri: fileToChange.resource,
+                    type: 'modify',
+                    state: 'pending',
+                    targetState: 'Hello World Modify!',
+                    changeSet,
+                    chatSessionId
+                })
+            );
+        }
+        if (fileToDelete && fileToDelete.resource) {
+            changeSet.addElement(
+                this.fileChangeFactory({
+                    uri: fileToDelete.resource,
+                    type: 'delete',
+                    state: 'pending',
+                    changeSet,
+                    chatSessionId
+                })
+            );
+        }
+        request.session.setChangeSet(changeSet);
+
+        request.response.response.addContent(new MarkdownChatResponseContentImpl(
+            'I have created a change set for you. You can now review and apply it.'
+        ));
+        request.response.complete();
```

Comment:
I wonder when we should update the change set in a real scenario. Simple solution would be when a response is there, but if more then one file is changed, it looks too late from a user POV. Unfortunately, the agent currently does not know which functions are called, therefore, we have no good trigger.

Reply:
As discussed, the agent can at any time update the chat model, and after https://github.com/eclipse-theia/theia/pull/14751 also tool functions can update it.